mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala

```diff
@@ -194,6 +232,285 @@ final class Word2Vec @Since("1.4.0") (
 
   @Since("1.4.1")
   override def copy(extra: ParamMap): Word2Vec = defaultCopy(extra)
+
+  /**
+   * Similar to InitUnigramTable in the original code. Instead of using an array of size 100 million
+   * like the original, we size it to be 20 times the vocabulary size.
+   * We sacrifice memory here, to get constant time lookups into this array when generating
+   * negative samples.
+   */
+  private def generateUnigramTable(normalizedWeights: Array[Double], tableSize: Int): Array[Int] = {
+    val table = Array.fill(tableSize)(0)
+    var a = 0
+    var i = 0
+    while (a < table.length) {
+      table.update(a, i)
+      if (a.toFloat / table.length >= normalizedWeights(i)) {
+        i = math.min(normalizedWeights.length - 1, i + 1)
+      }
+      a += 1
+    }
+    table
+  }
+
+  private def generateVocab[S <: Iterable[String]](input: RDD[S]):
+      (Int, Long, Map[String, Int], Array[Int]) = {
+    val sc = input.context
+
+    val words = input.flatMap(x => x)
+
+    val vocab = words.map(w => (w, 1L))
+      .reduceByKey(_ + _)
+      .filter{case (w, c) => c >= $(minCount)}
```

Comment:
You should broadcast ``$(minCount)`` to avoid shipping the entire class to each task. (and make sure to destroy it at the end of the method)